bluesky_browser/suitcase.py

```diff
@@ -0,0 +1,558 @@
+import traitlets
+from PyQt5.QtWidgets import (QPushButton, QVBoxLayout, QTabWidget, QCheckBox,
+                             QFormLayout, QLineEdit, QSpinBox)
+import suitcase.csv
+import suitcase.json_metadata
+import suitcase.specfile
+import suitcase.tiff_stack
+import suitcase.tiff_series
+
+from .utils import (ConfigurableQWidget, QStrListEdit, QIntListEdit, QBoolBox,
+                    QBoolNoneBox)
+
+
+class SuitcaseWidget(ConfigurableQWidget):
+    '''Widget for entering generic ``suitcase.xyz.export()`` parameters.
+
+    Parameters
+    ----------
+    *args : args
+        args passed down to QWidget
+    geometry : tuple
+        Tuple giving the (left, top, width, height) required for the window
+    **kwargs : kwargs
+        kwargs passed down to QWidget
+
+    Configurable Traits
+    -------------------
+    directory : str
+        The directory passed to the ``suitcase.xyz`` exporter
+    file_prefix : str
+        The file_prefix passed to the ``suitcase.xyz`` exporter
+    show : list
+        A list of parameters to include in the dialog box
+    '''
+
+    file_prefix = traitlets.Unicode('test_data_{start[scan_id]}')
+    directory = traitlets.Unicode('test_data')
+    parameters = traitlets.List(['directory', 'file_prefix'])
+
+    def __init__(self, *args, geometry=(10, 10, 400, 140), **kwargs):
+        super().__init__(*args, **kwargs)
+        self.geometry = geometry
+        self.initGUI()
+        self.show_parameters()
+
+    def initGUI(self):
+        self.setGeometry(*self.geometry)
+
+        self.main_layout = QFormLayout()
+
+        self.enable_checkbox = QCheckBox('enable exporter', self)
+        self.main_layout.addRow('', self.enable_checkbox)
+
+        # 'directory' trait
+        # NOTE: These four values should be updated for different traits.
+        trait_name = 'directory'  # The name of the trait
+        widget_type = QLineEdit  # The QT widget for this trait
+        slot_name = 'setText'  # The slot name for this trait
+        signal_name = 'textChanged'  # The signal name for this trait
+        # reference the trait and create the trait widget
```

Comment:
Have you considered making this into a method that takes the four variables above as arguments?

Reply:
@danielballan I have considered it, but am still a little on the fence.

I assume the method would be in the base ``SuitcaseWidget`` class and look something like.

```
def _create_trait_widget(self, trait_name, widget_type, slot_name, signal_name):
    # reference the trait and create the trait widget
    trait = getattr(self, trait_name)
    setattr(self, f'{trait_name}_widget', widget_type())
    trait_widget = getattr(self, f'{trait_name}_widget')
    # write the value of the trait to the widget
    if trait is not None:  # ignore if it is "None"
    getattr(trait_widget, slot_name)(trait)
    # Connect the output of the widget to the file_prefix trait.
    getattr(trait_widget, signal_name).connect(
        lambda output: setattr(self, trait_name, output))
```
And we would then do a series of definitions in each of the child classes like:
```
self._create_trait_widget('directory', QLineEdit, 'setText' 'textChanged' )
self._create_trait_widget('file_prefix', QLineEdit, 'setText' 'textChanged' )
...
```

I am not against this, it does improve readability. I suppose my only (not strongly held) concern is that we are starting to head toward conceptual complexity again.

Is this what you had in mind?

Reply:
Yes. I think the benefits justify the new abstraction in this case.

Reply:
OK I have made this change.